### ISSUE TITLE
Buttons doesn't work

### ISSUE THREAD

**Issue by cht47** [matterpoll/matterpoll#174]
#### Summary
Using the mattermost docker container. Everything else is working for years now. 
Mattermost version is 5.12.1, except some preview errors everything is ok, other plugins working fine.

#### Steps to reproduce
1. create Poll
2. Click on any button
3. nothing happens

#### Expected behavior
buttons work

#### Observed behavior (that appears unintentional)
poll doesn't work

#### Server configuration
Linux, Docker

**Mattermost version:**
 5.12.1

**Matterpoll version:**
1.1.0

**Updated from an older Matterpoll version or fresh install:**
fresh

<details>
<summary>Mattermost server logs</summary>

```
{"level":"error","ts":----,"caller":"mlog/log.go:172","msg":"Action integration error","path":"/api/v4/posts/o8hj6qanz3848rcgjouin35tfa/actions/umkr49ji7ido3mbtattx7xpa3a","request_id":"xxxxx,"ip_addr":"some lan ip","user_id":"h....","method":"POST","err_where":"DoActionRequest","http_code":400,"err_details":"err=Post https://some.domain:8181/plugins/com.github.matterpoll.matterpoll/api/v1/polls/rpug5s6ua7yd8xu9h716s8b7by/delete: dial tcp my.external.ip.address:8181: i/o timeout"}
```
</details>

#### Additional context
using the webcontainer which does https, site url is correctly set to https and port.
Webhooks, Hubot etc are working fine.

**Comment by cht47** [comment]
Searching through google, it seems to be a common problem with the docker container. My site URL is set to HTTPS (which is correctly) but the Mattermost instance is not https, its http and maybe that is where the problem starts. (Hubot is working with TLS)

**Comment by hanzei** [comment]
>  other plugins working fine.

What other plugins did you test?

> My site URL is set to HTTPS (which is correctly)

Are you using some proxy like nginx for the tls termination or the MM build in tool?

**Comment by cht47** [comment]
> > other plugins working fine.
> 
> What other plugins did you test?
> 
> > My site URL is set to HTTPS (which is correctly)
> 
> Are you using some proxy like nginx for the tls termination or the MM build in tool?

Plugins: Autolink and Giphy.

Its an Nginx which acts as reverse proxy and it handles the TLS. 
From the yaml file:
    ports:
      - "80:80"
      - "8181:443"

**Comment by hanzei** [comment]
> Plugins: Autolink and Giphy.

What Giphy plugin did you use exactly? There are at least two. Did the plugin work without a problem?

**Comment by cht47** [comment]
> > Plugins: Autolink and Giphy.
> 
> What Giphy plugin did you use exactly? There are at least two. Did the plugin work without a problem?

com.github.moussetc.mattermost.plugin.giphy - 0.2.0

**Comment by hanzei** [comment]
Would you please try to get another plugin with interactive buttons working like https://github.com/scottleedavis/mattermost-plugin-remind/?

**Comment by makhoff** [comment]
Hello, I have the same issue with buttons.
Server published with nginx.
`# HTTPS Server
    server {
        listen 443 ssl;
        server_name MYSERVER;
        error_log /var/log/nginx/mattermost_error.log;
        large_client_header_buffers 4 16k;

        ssl_certificate /etc/nginx/MYCERT;
        ssl_certificate_key /etc/nginx/MYCERT.key;
        ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
        ssl_ciphers 'ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA';
        ssl_prefer_server_ciphers on;
        ssl_session_cache shared:SSL:20m;
        ssl_session_timeout 180m;

        location / {
            proxy_pass http://MYSERVER:8065/;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
            proxy_set_header Host $http_host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forward-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forward-Proto http;
            proxy_set_header X-Nginx-Proxy true;
            proxy_redirect off;
        }
    }
`

**Comment by cht47** [comment]
> Would you please try to get another plugin with interactive buttons working like https://github.com/scottleedavis/mattermost-plugin-remind/?

Hello, sorry I was not in the office. I will try it tomorrow.

**Comment by makhoff** [comment]
Hi.

This is "reminder" plugin and it didn't applicable for voting(as I see in description).


>Tuesday, July 23, 2019 4:14 PM +03:00 from cht47 <notifications@github.com>:
>
>>Would you please try to get another plugin with interactive buttons working like  https://github.com/scottleedavis/mattermost-plugin-remind/ ?
>Hello, sorry I was not in the office. I will try it tomorrow.
>—
>You are receiving this because you commented.
>Reply to this email directly,  view it on GitHub , or  mute the thread .

M S

**Comment by cht47** [comment]
> Would you please try to get another plugin with interactive buttons working like https://github.com/scottleedavis/mattermost-plugin-remind/?

Ok I tried the plugin. It doesn't work either. -> Action integration error while clicking on the button. 

If makoff has the same problem its not docker related, the thing we have in common is a nginx ReverseProxy config.

**Comment by makhoff** [comment]
> If makoff has the same problem its not docker related, the thing we have in common is a nginx ReverseProxy config. So it seems its a Mattermost bug. Or do you have any other suggestions?

I think so. It's a nginx issue.

**Comment by kaakaa** [comment]
Is `dial tcp my.external.ip.address:8181: i/o timeout` always the cause of the error?
If so, it might indicate that dialing takes too long.

For confirming it, I want to know the duration between receipt of post action request and error occurrence.

Mattermost writes the debug log about  receipt of post action like the first line of the following example.

```
...
{"level":"debug","ts":1564013986.4115944,"caller":"web/handlers.go:58","msg":"POST - /api/v4/posts/o8hj6qanz3848rcgjouin35tfa/actions/umkr49ji7ido3mbtattx7xpa3a"}
...
{"level":"error","ts":1564013990.4115944,"caller":"mlog/log.go:172","msg":"Action integration error","path":"/api/v4/posts/o8hj6qanz3848rcgjouin35tfa/actions/umkr49ji7ido3mbtattx7xpa3a","request_id":"xxxxx,"ip_addr":"some lan ip","user_id":"h....","method":"POST","err_where":"DoActionRequest","http_code":400,"err_details":"err=Post https://some.domain:8181/plugins/com.github.matterpoll.matterpoll/api/v1/polls/rpug5s6ua7yd8xu9h716s8b7by/delete: dial tcp my.external.ip.address:8181: i/o timeout"}
...
```

Since dial timeout settings of Mattermost is [set to 3 seconds](https://github.com/mattermost/mattermost-server/blob/7ac5715a025128549243ca0f44436a263e57ebf6/services/httpservice/client.go#L140), if the duration between two logs is 3 seconds or more, it indicates that this issue is related to network environment.

---

I can reproduce this issue by setting the timeout (`ConnectTimeout` in [here](https://github.com/mattermost/mattermost-server/blob/7ac5715a025128549243ca0f44436a263e57ebf6/services/httpservice/client.go#L140)) to `1 * time.MilliSecond`.

```
2019-07-25T10:07:37.420+0900	debug	web/handlers.go:58	POST - /api/v4/posts/irfnu18pmbb3te3wfaoq3d8r1w/actions/8bxqpknyt7y5jez3ciiyx9arya
2019-07-25T10:07:37.432+0900	error	mlog/log.go:172	Action integration error	{"path": "/api/v4/posts/irfnu18pmbb3te3wfaoq3d8r1w/actions/8bxqpknyt7y5jez3ciiyx9arya", "request_id": "8tekrmdoxfyybgq9uie9wyj7yh", "ip_addr": "::1", "user_id": "xgqr5z7753bubrmdhegueximyr", "method": "POST", "err_where": "DoActionRequest", "http_code": 400, "err_details": "err=Post http://localhost:8065/plugins/com.github.matterpoll.matterpoll/api/v1/polls/bezhuif1i3yx3k6ry8krkb9fuh/vote/0: dial tcp: i/o timeout"}
```

**Comment by cht47** [comment]
I've changed the site url to mattermost directly (without nginx) and port 80... 

"method":"POST","err_where":"DoActionRequest","http_code":400,"err_details":"status=403"}

not a i/o timeout but still not working.

**Comment by hanzei** [comment]
> I think that feature is just buggy. Wouldn't be the first bug in this release. Docker Container is now on 5.13, lets try that.

Did upgrading your docker fix the issue?

**Comment by cht47** [comment]
> > I think that feature is just buggy. Wouldn't be the first bug in this release. Docker Container is now on 5.13, lets try that.
> 
> Did upgrading your docker fix the issue?

I'm currently on vacation and didn't want to upgrade and leave it without enough testing. I can do the upgrade probably next week.

**Comment by hanzei** [comment]
@cht47 Just wanted to get in touch again. Did you got the time to upgrade your system?

**Comment by cht47** [comment]
> @cht47 Just wanted to get in touch again. Did you got the time to upgrade your system?

Yes and still not working.

**Comment by guillaume-perreal** [comment]
Same issue here, using Apache as a reverse-proxy and TLS all along. The site URL is properly set. Other features work fine. I've got the same errors.

Log message:
```
{"level":"error","ts":1568713536.304262,"caller":"mlog/log.go:173","msg":"Action integration error","path":"/api/v4/posts/aathm7oqj7rwfqs8jbtiiajz8c/actions/ipfsuacaifym3npskkc37yysph","request_id":"4huyuozcn78y9crdfpufm1goua","ip_addr":"[REDACTED CLIENT IP ADDRESS]","user_id":"ahjk8ubtub865jd4gdg64ekiua","method":"POST","err_where":"DoActionRequest","http_code":400,"err_details":"err=Post https://[REDACTED PUBLIC HOSTNAME]/plugins/com.github.matterpoll.matterpoll/api/v1/polls/wxteeoqma3nyxm6hqua7bkxdtr/vote/0: dial tcp [REDACTED REVERSE-PROXY IP ADDRESS]: i/o timeout"}
```

**Comment by blu3r4y** [comment]
I am also using Mattermost inside a Docker container and behind an NGINX reverse proxy, where HTTPS is terminated by the proxy already. I got the errors `no route to host` and `i/o timeout`.

The problem is that the matterpoll plugin tries to reach `https://your-mattermost-instance.com`, which is (usually) not reachable from inside the Docker container.

I resolved this by using the local Docker DNS to redirect those request directly to the NGINX container. I used the [legacy `--link` option](https://docs.docker.com/network/links/) because it is convenient. Maybe think about using user-defined networks instead.

Assuming that your NGINX container is named `nginx`, you can start your Mattermost container like so:

```
docker run
  --link nginx:your-mattermost-instance
  # ... your arguments ...
  mattermost/mattermost-prod-app:latest
```

This will redirect HTTPS queries `https://your-mattermost-instance.com` directly to your reverse proxy `https://nginx` instead.

**Comment by satochan0112** [comment]
I had a same problem.
I don't use docker, but I use a load balancer to transfer to the aws ec2 mattermost server. Since all SSL certificates are attached to the load balancer, https is used for external access, but http is used inside ec2.

ALB's IP is returned when i send following command on ec2 mm server, but it cannot be accessed from the server. 

`nslookup my.mattermost.domain`

(Despite the path goal being the instance itself!) 
To that proof:

`ping my.mattermost.domain`

is no response. So, i edit the hosts file and route it.

`127.0.0.1 localhost localhost.localdomain localhost4 localhost4.localdomain4 my.mattermost.domain`

It worked.

**Comment by hanzei** [comment]
Closing this one as it's very docker specific and not limited to this plugin. 

https://github.com/mattermost/mattermost-server/pull/12156 will fix the underlying problem anyway in MM 5.18.
